### ISSUE TITLE
[BUG] azure-messaging-servicebus:7.15.1 -  Application not consume any message from Azure Service Bus after logs "The connection was closed by container  because it did not have any active links in the past 300000 milliseconds" printed on AKS

### ISSUE THREAD

**Issue by nguyendangtu** [Azure/azure-sdk-for-java#41489]
**Describe the bug**
The application working well with Spring, Tomcat, and azure-messaging-servicebus:7.10.1.
While migrate from 7.10.1 to 7.15.1, application working well on local (macbook) but not working well on Azure Kubernetes Service. The Connection closed by container as ServiceBusReceiveLinkProcessor not detected action links during 300000 milliseconds.

***Exception or Stack Trace***
2024/08/05 20:40:07.477 [reactor-executor-14] WARN  c.a.c.a.i.RequestResponseChannel - {"az.sdk.message":"Error in SendLinkHandler. Disposing unconfirmed sends.","exception":"The connection was closed by container 'xxx_G18' because it did not have any active links in the past 300000 milliseconds. TrackingId:xxx_G18, SystemTracker:gateway5, Timestamp:2024-08-05T20:40:07, errorContext[NAMESPACE: <name-space>.servicebus.windows.net. ERROR CONTEXT: N/A, PATH: $cbs, REFERENCE_ID: cbs:sender, LINK_CREDIT: 99]","connectionId":"yyy","linkName":"cbs"}

2024/08/05 20:40:16.567 [receiverPump-4] WARN  c.a.c.a.implementation.MessageFlux - {"az.sdk.message":"Receiver emitted terminal error.","exception":"The link 'G13:122632033:<subscription_name>/subscriptions/zzz' is force detached. Code: consumer(link70350). Details: AmqpMessageConsumer.IdleTimerExpired: Idle timeout: 00:10:00. TrackingId:xxx_G13_B18, SystemTracker:gi::G13:122632033:<subcription_name>/subscriptions/zzzz, bi::in-connection772(G13-3498347)::session773::link70350, Timestamp:2024-08-05T20:40:16, errorContext[NAMESPACE: <namespace>.servicebus.windows.net. ERROR CONTEXT: N/A, PATH: <subscription_name>/subscriptions/yyy, REFERENCE_ID: <subcription_name>/subscriptions/yyy, LINK_CREDIT: 0]","messageFlux":"mf_8ddb3a_1722886936377","connectionId":"MF_636c89_1722886936378","linkName":"<subcription_name>/subscriptions/<topic_name>_ea7fa2_1722886936377","entityPath":"<subcription_name>/subscriptions/<topic_name>"}
2024/08/05 20:40:16.568 [receiverPump-4] WARN  c.a.c.a.implementation.MessageFlux - {"az.sdk.message":"Current mediator reached terminal error-state (retriable:true).","exception":"The link 'G13:122632033:<subcription_name>/subscriptions/<topic_name>_ea7fa2_1722886936377' is force detached. Code: consumer(link70350). Details: AmqpMessageConsumer.IdleTimerExpired: Idle timeout: 00:10:00. TrackingId:f70403d60017141c000112ce66b12b1a_G13_B18, SystemTracker:gi::G13:122632033:<subcription_name>/subscriptions/<topic_name>_ea7fa2_1722886936377, bi::in-connection772(G13-3498347)::session773::link70350, Timestamp:2024-08-05T20:40:16, errorContext[NAMESPACE: <name_space>.servicebus.windows.net. ERROR CONTEXT: N/A, PATH: <subcription_name>/subscriptions/<topic_name>, REFERENCE_ID: <subcription_name>/subscriptions/<topic_name>_ea7fa2_1722886936377, LINK_CREDIT: 0]","messageFlux":"mf_8ddb3a_1722886936377","connectionId":"MF_636c89_1722886936378","linkName":"<subcription_name>/subscriptions/<topic_name>_ea7fa2_1722886936377","entityPath":"<subcription_name>/subscriptions/<topic_name>","attempt":1,"retryAfter":4511}
2024/08/05 20:40:21.083 [parallel-4] WARN  c.a.c.a.implementation.MessageFlux - {"az.sdk.message":"Requesting a new mediator.","messageFlux":"mf_8ddb3a_1722886936377","connectionId":"MF_636c89_1722886936378","linkName":"<subcription_name>/subscriptions/<topic_name>_ea7fa2_1722886936377","entityPath":"<subcription_name>/subscriptions/<topic_name>"}
2024/08/05 20:40:21.125 [reactor-executor-7] WARN  c.a.c.a.implementation.MessageFlux - {"az.sdk.message":"Setting next mediator and waiting for activation.","messageFlux":"mf_8ddb3a_1722886936377","connectionId":"MF_636c89_1722886936378","linkName":"<subcription_name>/subscriptions/<topic_name>_ea7fa2_1722886936377","entityPath":"<subcription_name>/subscriptions/<topic_name>"}
2024/08/05 20:40:21.164 [receiverPump-13] WARN  c.a.c.a.implementation.MessageFlux - {"az.sdk.message":"The mediator is active.","messageFlux":"mf_8ddb3a_1722886936377","connectionId":"MF_636c89_1722886936378","linkName":"<subcription_name>/subscriptions/<topic_name>_ea7fa2_1722886936377","entityPath":"<subcription_name>/subscriptions/<topic_name>"}

2024/08/05 20:41:01.390 [receiverPump-7] WARN  c.a.c.a.implementation.MessageFlux - {"az.sdk.message":"Receiver emitted terminal error.","exception":"The link 'G20:60205512:<subcription_name>/subscriptions/delete_d4fbe7_1722886935378' is force detached. Code: consumer(link70356). Details: AmqpMessageConsumer.IdleTimerExpired: Idle timeout: 00:10:00. TrackingId:f70403d60005141c000112d466b12b1a_G20_B18, SystemTracker:gi::G20:60205512:<subcription_name>/subscriptions/delete_d4fbe7_1722886935378, bi::in-connection6011(G20-195874)::session6012::link70356, Timestamp:2024-08-05T20:41:01, errorContext[NAMESPACE: <name_space>.servicebus.windows.net. ERROR CONTEXT: N/A, PATH: <subcription_name>/subscriptions/delete, REFERENCE_ID: <subcription_name>/subscriptions/delete_d4fbe7_1722886935378, LINK_CREDIT: 0]","messageFlux":"mf_72cff5_1722886935378","connectionId":"MF_60b1a3_1722886935379","linkName":"<subcription_name>/subscriptions/delete_d4fbe7_1722886935378","entityPath":"<subcription_name>/subscriptions/delete"}

**To Reproduce**
Step 1: Setup a application with Spring WebFlux (Spring 5) running on tomcat
Step 2: Setup a AKS environment to deploy the service to 3 regions (EUS, SCUS, WUS). min is 4, and max is 10 pods.
Step 3: Create a 10 subscriptions, each subscription  have 10 topic to subscribe message, min delivery count is 3, and max delivery count is 10

***Code Snippet***
```
			ConnectionStringProperties properties = new ConnectionStringProperties(connectionString);

			this.subscriptionClient = new ServiceBusClientBuilder()
					.connectionString(connectionString)
					.processor()
					.topicName("ABC")
					.subscriptionName("DEF")
					.processMessage(this::processMessage)
					.processError(this::processError)
					.receiveMode(ServiceBusReceiveMode.PEEK_LOCK)
					.maxAutoLockRenewDuration(Duration.ofMinutes(config.getMaxAutoRenewDuration()))
					.maxConcurrentCalls(config.getMaxConcurrentCalls())
					.disableAutoComplete()
					.buildProcessorClient();

			this.subscriptionClient.start();
```

**Expected behavior**
Expected that no code change and version 7.15.1 is working well on AKS as version 7.10.1

**Screenshots**
No screenshot, service on AKS not able to consumer message from azure service bus. Already provide logs

**Setup (please complete the following information):**
 - Environment: [Azure Kubernetes Service]
 - Library/Libraries: [com.azure:azure-core:1.46.0, azure-messaging-servicebus:jar:7.15.1], 
 - Java version: [11] 
 - App Server/Environment: [Tomcat]
 - Frameworks: [Spring 5] 

Libs:
[INFO] |  +- com.azure:azure-messaging-servicebus:jar:7.15.1:compile
[INFO] |  |  +- com.azure:azure-core:jar:1.46.0:compile
[INFO] |  |  |  +- com.azure:azure-json:jar:1.1.0:compile
[INFO] |  |  |  \- com.fasterxml.jackson.datatype:jackson-datatype-jsr310:jar:2.13.3:compile
[INFO] |  |  +- com.azure:azure-core-amqp:jar:2.9.1:compile
[INFO] |  |  |  +- com.microsoft.azure:qpid-proton-j-extensions:jar:1.2.4:compile
[INFO] |  |  |  \- org.apache.qpid:proton-j:jar:0.33.8:compile
[INFO] |  |  +- com.azure:azure-core-http-netty:jar:1.14.0:compile
[INFO] |  |  |  +- io.netty:netty-handler:jar:4.1.101.Final:compile
[INFO] |  |  |  |  +- io.netty:netty-resolver:jar:4.1.101.Final:compile
[INFO] |  |  |  |  \- io.netty:netty-transport:jar:4.1.101.Final:compile
[INFO] |  |  |  +- io.netty:netty-handler-proxy:jar:4.1.101.Final:compile
[INFO] |  |  |  |  \- io.netty:netty-codec-socks:jar:4.1.101.Final:compile
[INFO] |  |  |  +- io.netty:netty-buffer:jar:4.1.101.Final:compile
[INFO] |  |  |  +- io.netty:netty-codec:jar:4.1.101.Final:compile
[INFO] |  |  |  +- io.netty:netty-codec-http:jar:4.1.101.Final:compile
[INFO] |  |  |  +- io.netty:netty-codec-http2:jar:4.1.101.Final:compile
[INFO] |  |  |  +- io.netty:netty-transport-native-unix-common:jar:4.1.101.Final:compile
[INFO] |  |  |  +- io.netty:netty-transport-native-epoll:jar:linux-x86_64:4.1.101.Final:compile
[INFO] |  |  |  |  \- io.netty:netty-transport-classes-epoll:jar:4.1.101.Final:compile
[INFO] |  |  |  +- io.netty:netty-transport-native-kqueue:jar:osx-x86_64:4.1.101.Final:compile
[INFO] |  |  |  |  \- io.netty:netty-transport-classes-kqueue:jar:4.1.101.Final:compile
[INFO] |  |  |  +- io.netty:netty-tcnative-boringssl-static:jar:2.0.62.Final:compile
[INFO] |  |  |  |  +- io.netty:netty-tcnative-classes:jar:2.0.62.Final:compile
[INFO] |  |  |  |  +- io.netty:netty-tcnative-boringssl-static:jar:linux-x86_64:2.0.62.Final:compile
[INFO] |  |  |  |  +- io.netty:netty-tcnative-boringssl-static:jar:linux-aarch_64:2.0.62.Final:compile
[INFO] |  |  |  |  +- io.netty:netty-tcnative-boringssl-static:jar:osx-x86_64:2.0.62.Final:compile
[INFO] |  |  |  |  +- io.netty:netty-tcnative-boringssl-static:jar:osx-aarch_64:2.0.62.Final:compile
[INFO] |  |  |  |  \- io.netty:netty-tcnative-boringssl-static:jar:windows-x86_64:2.0.62.Final:compile
[INFO] |  |  |  +- io.projectreactor.netty:reactor-netty-http:jar:1.0.40:compile
[INFO] |  |  |  |  +- io.netty:netty-resolver-dns:jar:4.1.101.Final:compile
[INFO] |  |  |  |  |  \- io.netty:netty-codec-dns:jar:4.1.101.Final:compile
[INFO] |  |  |  |  +- io.netty:netty-resolver-dns-native-macos:jar:osx-x86_64:4.1.101.Final:compile
[INFO] |  |  |  |  |  \- io.netty:netty-resolver-dns-classes-macos:jar:4.1.101.Final:compile
[INFO] |  |  |  |  \- io.projectreactor.netty:reactor-netty-core:jar:1.0.40:compile
[INFO] |  |  |  \- io.netty:netty-common:jar:4.1.101.Final:compile

Glad to looking forward to hear from you soon.

Thanks,
John

**Comment by nguyendangtu** [comment]
I add more stack trace:

```
2024/08/11 14:30:28.776 [boundedElastic-10] ERROR c.a.m.s.ServiceBusReceiverAsyncClient - Cannot perform operation 'abandoned' on a disposed receiver.
2024-08-11T14:30:28.883 WARNING org.apache.catalina.loader.WebappClassLoaderBase The web application [ROOT] appears to have started a thread named [reactor-executor-1] but has failed to stop it. This is very likely to create a memory leak. Stack trace of thread:
 java.base@11.0.20.1.101/sun.nio.ch.EPoll.wait(Native Method)
 java.base@11.0.20.1.101/sun.nio.ch.EPollSelectorImpl.doSelect(Unknown Source)
 java.base@11.0.20.1.101/sun.nio.ch.SelectorImpl.lockAndDoSelect(Unknown Source)
 java.base@11.0.20.1.101/sun.nio.ch.SelectorImpl.select(Unknown Source)
 org.apache.qpid.proton.reactor.impl.SelectorImpl.select(SelectorImpl.java:126)
 org.apache.qpid.proton.reactor.impl.IOHandler.handleQuiesced(IOHandler.java:57)
 org.apache.qpid.proton.reactor.impl.IOHandler.onUnhandled(IOHandler.java:390)
 org.apache.qpid.proton.engine.BaseHandler.onReactorQuiesced(BaseHandler.java:87)
 org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:206)
 org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
 org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
 org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:292)
 com.azure.core.amqp.implementation.ReactorExecutor.run(ReactorExecutor.java:91)
 com.azure.core.amqp.implementation.ReactorExecutor$$Lambda$1467/0x0000000100f03040.run(Unknown Source)
 reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
 reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
 java.base@11.0.20.1.101/java.util.concurrent.FutureTask.run(Unknown Source)
 java.base@11.0.20.1.101/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
 java.base@11.0.20.1.101/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
 java.base@11.0.20.1.101/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
 java.base@11.0.20.1.101/java.lang.Thread.run(Unknown Source)

```

**Comment by conniey** [comment]
>  ...appears to have started a thread named [reactor-executor-1] ...

We keep the reactor-executor alive for a period of time (tryTimeout duration) to drain any in-flight calls.  After the tryTimeout duration elapses, the reactor-executor-1 scheduler should be closed.

**Comment by conniey** [comment]
* Since 7.15.1, we have made several reliability fixes.  Does this reproduce in azure-messaging-servicebus 7.17.2?
* If it does, do you have a repro/set-up instructions we can execute?
* Do you have INFO/DEBUG logs within +/- minutes we can look at?

**Comment by nguyendangtu** [comment]
Hi team,

I have found that to troubleshoot the azure service bus issue from https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/servicebus/azure-messaging-servicebus/TROUBLESHOOTING.md.
My application stop consume message from Azure Service Bus because default pool size 2 * CPUs were not support.
It is working well after I reduce number of Concurrent calls.
Could you please help to add logs for this issue to help developers identify and knowing the issue while application stop consume messaging.

Thanks,
John

**Comment by djarnis73** [comment]
I'm seeing similar error messages in our logs this as well. It started Aug 19. around noon on an application that has not been changed for a while, so my suspicion was a change in azure infrastructure.
 We are using `com.azure:azure-messaging-eventhubs:5.18.6`

```
reactor.core.Exceptions$ErrorCallbackNotImplemented: com.azure.core.amqp.exception.AmqpException: onSessionRemoteClose connectionId[MF_81d2b1_1723864501179], entityName[cbs-session] condition[Error{condition=amqp:connection:forced, description='The connection was closed by container 'a33e5dd3f7214537b15e7d4aae81aea4_G25' because it did not have any active links in the past 300000 milliseconds. TrackingId:a33e5dd3f7214537b15e7d4aae81aea4_G25, SystemTracker:gateway5, Timestamp:2024-08-17T10:40:01', info=null}], errorContext[NAMESPACE: evh-bibis-prod-001.servicebus.windows.net. ERROR CONTEXT: N/A, PATH: cbs-session]
Caused by: com.azure.core.amqp.exception.AmqpException: onSessionRemoteClose connectionId[MF_81d2b1_1723864501179], entityName[cbs-session] condition[Error{condition=amqp:connection:forced, description='The connection was closed by container 'a33e5dd3f7214537b15e7d4aae81aea4_G25' because it did not have any active links in the past 300000 milliseconds. TrackingId:a33e5dd3f7214537b15e7d4aae81aea4_G25, SystemTracker:gateway5, Timestamp:2024-08-17T10:40:01', info=null}], errorContext[NAMESPACE: evh-bibis-prod-001.servicebus.windows.net. ERROR CONTEXT: N/A, PATH: cbs-session]
	at com.azure.core.amqp.implementation.ExceptionUtil.toException(ExceptionUtil.java:90)
	at com.azure.core.amqp.implementation.handler.SessionHandler.onSessionRemoteClose(SessionHandler.java:139)
	at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:152)
	at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
	at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
	at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:291)
	at com.azure.core.amqp.implementation.ReactorExecutor.run(ReactorExecutor.java:91)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```

It also seems to trigger some internal NPE, because I'm also seeing this exception at the same time:

```
reactor.core.Exceptions$ErrorCallbackNotImplemented: java.lang.NullPointerException: Cannot invoke "java.util.List.add(Object)" because "this._sessions" is null
Caused by: java.lang.NullPointerException: Cannot invoke "java.util.List.add(Object)" because "this._sessions" is null
	at org.apache.qpid.proton.engine.impl.ConnectionImpl.session(ConnectionImpl.java:91)
	at org.apache.qpid.proton.engine.impl.ConnectionImpl.session(ConnectionImpl.java:39)
	at com.azure.core.amqp.implementation.ReactorConnection.lambda$createSession$15(ReactorConnection.java:311)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(Unknown Source)
	at com.azure.core.amqp.implementation.ReactorConnection.lambda$createSession$16(ReactorConnection.java:308)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:113)
	at reactor.core.publisher.MonoPeekTerminal$MonoTerminalPeekSubscriber.onNext(MonoPeekTerminal.java:180)
	at reactor.core.publisher.FluxHide$SuppressFuseableSubscriber.onNext(FluxHide.java:137)
	at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.request(MonoIgnoreThen.java:164)
	at reactor.core.publisher.FluxHide$SuppressFuseableSubscriber.request(FluxHide.java:152)
	at reactor.core.publisher.MonoPeekTerminal$MonoTerminalPeekSubscriber.request(MonoPeekTerminal.java:139)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:171)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.request(MonoFlatMap.java:194)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:171)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:171)
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.request(FluxPeekFuseable.java:144)
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.request(Operators.java:2331)
	at com.azure.core.amqp.implementation.AmqpChannelProcessor.requestUpstream(AmqpChannelProcessor.java:336)
	at com.azure.core.amqp.implementation.AmqpChannelProcessor.lambda$onError$4(AmqpChannelProcessor.java:230)
	at reactor.core.publisher.LambdaMonoSubscriber.onNext(LambdaMonoSubscriber.java:171)
	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.propagateDelay(MonoDelay.java:270)
	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:285)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)

```

**Comment by anuchandy** [comment]
Hello @nguyendangtu, Great to hear that adjusting the thread pool size resolved issues in your environment.

We've tried to replicate this issue by limiting the cores in both local Docker and AKS environments but haven't been successful. We've pushed a ready-to-deploy application here: https://github.com/anuchandy/servicebus-concurrentcalls-core.  This has a README file with all instructions to deploy it, if you wish to check/try.

I suspect your setup includes other components affecting threading, which limits CPU cycles for the Service Bus library when it needs to recover from network disconnects. This could be due to Reactor components from Spring sharing the same thread pool, or API/DB calls in the Processor handler that also use this pool alongside other application logic, or something unknown.

Since this issue was consist in your application case, would you be able to extract out a minimal runnable/deployable project that reproduce this issue? If you can push it to your Git like I did above, we can try pull and debug this. Based on the root causing, we can decide if it should be something addressed in Service Bus library or should be a detailed section in trouble shooting guideline.

**Comment by nguyendangtu** [comment]
> Hello @nguyendangtu, Great to hear that adjusting the thread pool size resolved issues in your environment.
> 
> We've tried to replicate this issue by limiting the cores in both local Docker and AKS environments but haven't been successful. We've pushed a ready-to-deploy application here: https://github.com/anuchandy/servicebus-concurrentcalls-cores. This has a README file with all instructions to deploy it, if you wish to check/try.
> 
> I suspect your setup includes other components affecting threading, which limits CPU cycles for the Service Bus library when it needs to recover from network disconnects. This could be due to Reactor components from Spring sharing the same thread pool, or API/DB calls in the Processor handler that also use this pool (maybe implicitly) alongside other application logic, or something unknown.
> 
> Since this issue was consist in your application case, would you be able to extract out a minimal runnable/deployable project that reproduce this issue? If you can push it to your Git like I did above, we can try pull and debug this. Based on the root causing, we can decide if it should be something addressed in Service Bus library or should be a detailed section in trouble shooting guideline.

dear @anuchandy, My service have 10 pods on AKS, each pod have 2 CPUs and 2GB memory. Max heap size for each pod is 1GB. The number of concurrent calls configured is 45. I have 10 topic. so each pod will consume 450 concurrent. It is not issue when I started the service. The issue only happened after I trigger performance testing, the number of pods scale from 3 to 10, then it stop consume any messages from Azure Service Bus.

**Comment by anuchandy** [comment]
Thanks @nguyendangtu. That's the calculation I also understood from your initial comments i.e., each pod listens to one subscription in 10 topics with a concurrency of 45. So, a total concurrency of 10 * 45 = 450 in one pod with 2 CPUs.

I think your additional point regarding performance testing supports my earlier suspicion on why this happened. What likely happened was - under performance testing, some of the external (external to Service Bus library) Reactor components (for example Web Flux) or application logic indirectly started to heavily use or reserve the Threads in the Reactor thread pool, the same thread pool that Service Bus library also uses. When pool is in such state, the Service Bus library couldn't acquire a thread to run its recovery process. I believe just one pod (without any scaling) and forcing the app components to overuse all the threads in the pool are sufficient to reproduce this problem. 

I’m unsure why you didn't see this issue with earlier version, possibly the Reactor library grow the Thread Pool in a different way in older version. Though I do remember pool sizing / low cores causing hang even on older versions as well for some users. Unfortunately, it's hard to detect such global Thread Pool bottlenecks at runtime. We've some information on client patterns that can cause threading issues in troubleshoot guidelines https://learn.microsoft.com/en-us/azure/developer/java/sdk/troubleshooting-messaging-service-bus-overview  

I'm unsure if you have cycles to prepare a minimal project, otherwise with the given data points we can conclude that the root cause is what described above.

**Comment by nguyendangtu** [comment]
> Thanks @nguyendangtu. That's the calculation I also understood from your initial comments i.e., each pod listens to one subscription in 10 topics with a concurrency of 45. So, a total concurrency of 10 * 45 = 450 in one pod with 2 CPUs.
> 
> I think your additional point regarding performance testing supports my earlier suspicion on why this happened. What likely happened was - under performance testing, some of the external (external to Service Bus library) Reactor components (for example Web Flux) or application logic indirectly started to heavily use or reserve the Threads in the Reactor thread pool, the same thread pool that Service Bus library also uses. When pool is in such state, the Service Bus library couldn't acquire a thread to run its recovery process. I believe just one pod (without any scaling) and forcing the app components to overuse all the threads in the pool are sufficient to reproduce this problem.
> 
> I’m unsure why you didn't see this issue with earlier version, possibly the Reactor library grow the Thread Pool in a different way in older version. Though I do remember pool sizing / low cores causing hang even on older versions as well for some users. Unfortunately, it's hard to detect such global Thread Pool bottlenecks at runtime. We've some information on client patterns that can cause threading issues in troubleshoot guidelines https://learn.microsoft.com/en-us/azure/developer/java/sdk/troubleshooting-messaging-service-bus-overview
> 
> I'm unsure if you have cycles to prepare a minimal project, otherwise with the given data points we can conclude that the root cause is what described above.

dear @anuchandy, would you please help to add correct logs for azure messaging service bus. The problem because of pool size and leak resource allocated while logs printed "Container terminate connection". It leading me to the wrong way to investigate the issue while I reading troubleshoot document from day 1. 
Seem it is a critical issue for concurrent calls, can you help to add hand-on labs for concurrent calls. It is recommendation 20-30 for pool size while the message consuming mostly base on memory, HPA also scale up base on memory too. May I know which is the formula between reactor pool size and number of pool size configuration per CPU?

**Comment by anuchandy** [comment]
Hi @nguyendangtu, the projectreactor.io documentation has a section describing the formula that Reactor library uses to compute the pool size from the cores, here is link to their doc [projectreactor.io](https://projectreactor.io/docs/core/release/api/reactor/core/scheduler/Schedulers.html#DEFAULT_BOUNDED_ELASTIC_SIZE)

**Comment by anuchandy** [comment]
We've [added](https://github.com/Azure/azure-sdk-for-java/pull/41627) a runtime check for cores/pools/concurrency and log a _recommendation_ message linking to the troubleshooting guideline. This should be available in September release (either 7.17.4 or 7.18.*).
